store/download_test.go

```diff
@@ -435,20 +441,136 @@ func (s *downloadSuite) TestUseDeltas(c *C) {
 	}
 
 	for _, scenario := range scenarios {
-		if scenario.exeInCore {
-			osutil.CopyFile("/bin/true", exeInCorePath, 0)
-		} else {
-			os.Remove(exeInCorePath)
-		}
+		var hostXdelta3Cmd, coreInterpCmd *testutil.MockCmd
+
+		var cleanups []func()
+
+		comment := Commentf("%#v", scenario)
+
+		// setup the env var for the scenario
 		os.Setenv("SNAPD_USE_DELTAS_EXPERIMENTAL", scenario.env)
 		release.MockOnClassic(scenario.classic)
+
+		// setup binaries for the scenario
+		if scenario.exeInCore {
+			// We need both the xdelta3 command for determining the interpreter
+			// as well as the actual interpreter for executing the basic
+			// "xdelta3 config" command.
+			// For the interpreter, since that's how we execute xdelta3, mock
+			// that as a command, but we don't need to mock the xdelta3 command
+			// in the core snap since that doesn't get executed by our fake
+			// interpreter. Mocking the interpreter and executing that as a
+			// MockCommand has the advantage that it avoids the specific ELF
+			// handling that is per-arch, etc. of the real CommandFromSystemSnap
+			// implementation.
+
+			coreInterpCmd = testutil.MockCommand(c, interpInCorePath, "")
+
+			r := store.MockSnapdtoolCommandFromSystemSnap(func(name string, args ...string) (*exec.Cmd, error) {
+				c.Assert(name, Equals, "/usr/bin/xdelta3")
+				c.Assert(args, DeepEquals, []string{"config"})
+
+				interpArgs := append([]string{"--library-path", "/some/dir/from/etc/ld.so", exeInCorePath}, args...)
+				return exec.Command(coreInterpCmd.Exe(), interpArgs...), nil
+			})
+			cleanups = append(cleanups, r)
+
+			// Forget the calls to the interpreter at the end of the test - this
+			// deletes the log which otherwise would  continue to persist for
+			// each iteration leading to incorrect checks for the calls to the
+			// absolute binary that we mocked here, as the log file will be the
+			// same for each iteration.
+			// For the inverse reason, we don't need to forget calls for the
+			// hostXdelta3Cmd mock command, it gets a new dir with a new log
+			// file each iteration.
+			cleanups = append(cleanups, func() {
+				coreInterpCmd.ForgetCalls()
+			})
+		}
+
 		if scenario.exeInHost {
-			os.Setenv("PATH", origPath)
-		} else {
-			os.Setenv("PATH", altPath)
+			// just mock the xdelta3 command directly
+			hostXdelta3Cmd = testutil.MockCommand(c, "xdelta3", "")
```

Comment:
and this, to cleanup ?

Reply:
we do call Restore() on this one, but we call it before the second UseDeltas() invocation, in order to test the caching properly, hence it is not in a cleanup

Reply:
I added a comment to explain where the Restore() is located for the casual reader